rotkehlchen/externalapis/covalent.py

```diff
@@ -0,0 +1,260 @@
+            action: str,
+            address: str = None,
+            options: Optional[Dict[str, Any]] = None,
+            timeout: Optional[Tuple[int, int]] = None,
+    ) -> Optional[Dict[str, Any]]:
+        """Queries Covalent
+
+        May raise:
+        - RemoteError if there are any problems with reaching Covalent or if
+        an unexpected response is returned
+        """
+        query_str = f'https://api.covalenthq.com/v1/{self.chain_id}/{action}'
+        if address:
+            query_str += f'/{address}'
+        query_str += f'/{module}/'
+
+        query_str += f'?key={KEY}'
+
+        if options:
+            for name, value in options.items():
+                query_str += f'&{name}={value}'
+
+        retry = 0
+        while retry <= CONST_RETRY:
+            logger.debug(f'Querying covalent: {query_str}')  # noqa: E501 lgtm [py/clear-text-logging-sensitive-data]
+            try:
+                response = self.session.get(query_str, timeout=timeout if timeout else DEFAULT_TIMEOUT_TUPLE)  # noqa: E501
+            except requests.exceptions.RequestException as e:
+                # In timeout retry
+                if isinstance(e, requests.exceptions.Timeout):
+                    if retry == CONST_RETRY:
+                        return None
+                    retry += 1
+                    continue
+                raise RemoteError(f'Covalent API request failed due to {str(e)}') from e
+
+            result = response.json()
+
+            if response.status_code != 200:
+                if 'error_message' in result:
+                    error_message = result['error_message']
+                else:
+                    error_message = None
+                raise RemoteError(
+                    f'Covalent API request {response.url} failed '
+                    f'with HTTP status code {response.status_code} and '
+                    f'Error message: {error_message} and text',
+                    f'{response.text}',
+                )
+
+            # success, break out of the loop and return result
+            break
+        return result
+
+    def get_transactions(
+            self,
+            account: ChecksumEthAddress,
+            from_ts: Optional[Timestamp] = None,
+            to_ts: Optional[Timestamp] = None,
+    ) -> Optional[List[CovalentTransaction]]:
+        """Gets a list of transactions for account.
+        - account is address for wallet.
+        - to_ts is latest date.
+        - from_ts is oldest date.
+        - get_log is parameter for get logs in transactions.
+        May raise:
+        - RemoteError due to self._query(). Also if the returned result
+        is not in the expected format
+        """
+
+        if from_ts is None:
+            from_ts = Timestamp(0)
+        if to_ts is None:
+            to_ts = ts_now()
+
+        result_master = []
+        options = {'limit': COVALENT_QUERY_LIMIT, 'page-size': PAGESIZE}
+        for i in range(1, 7):
```

Comment:
Why 1 to 7 here? I am a bit lost. You seem to be doing the query 6 times. So what is that for exactly? I assume some kind of pagination, but:

1. Where/when do you stop if there is no more results for the time range?
2. What if there is more results than what you are asking for in the time range?

Reply:
> Okay so few more comments, but a lot less this time.
> 
> This PR does not deal with actually saving the Avalanche transactions in the DB or showing them in the UI, right?

Yes, we just made the endpoint for now

Reply:
> Why 1 to 7 here? I am a bit lost. You seem to be doing the query 6 times. So what is that for exactly? I assume some kind of pagination, but:
> 
> 1. Where/when do you stop if there is no more results for the time range?
> 2. What if there is more results than what you are asking for in the time range?

We fixed this by doing pagination with the conditional of time (if it has from_timestamp) or goes until there's nothing more to query.

Reply:
All right this seems like the correct approach for paginating their queries :+1: